Reader does not respond post Segment container migration.

**Problem description**
It was observed that the Controller's AbortEvent processor is not able to read events from _abortStream when the Segment container migration happen. (This issue was observed during the ControllerFailover system test run.) 
On further investigation it was observed that the `reader.readNextEvent(timeout)` is blocked for more than 10 mins i.e the AbortEvent was never read from the _abortStream.

The logs found during the current investigation are listed below.

1) Test logs: test aborts the transaction.
```2018-01-23 06:40:04,075 9802 [Time-limited test] INFO  i.p.t.system.ControllerFailoverTest - Trying to abort transaction 7a16b5ca-14b4-4f53-8fe0-a2f2057e0e77, by sending request to controller at tcp://172.16.18.21:9092```


2)Controller writes to abortStream.
```scrambler-slave-1-container-2/controller.log:2018-01-23 06:40:04,133 2922513 [controllerpool-3] DEBUG i.p.c.t.S.StreamTransactionMetadataTasks - Transaction 7a16b5ca-14b4-4f53-8fe0-a2f2057e0e77, sent request to _abortStream
scrambler-slave-1-container-2/controller.log:2018-01-23 06:42:09,594 3047974 [controllerpool-63] DEBUG i.p.c.t.S.StreamTransactionMetadataTasks - Transaction 7a16b5ca-14b4-4f53-8fe0-a2f2057e0e77, sent request to _abortStream
```
3) Controller reads the abort stream event @ 06:57 17 mins after the event was written.
```scrambler-slave-1-container-2/controller_2018-01-23.21.log:2018-01-23 06:57:27,101 3965481 [controllerpool-59] DEBUG i.p.c.s.e.r.AbortRequestHandler - Aborting transaction 7a16b5ca-14b4-4f53-8fe0-a2f2057e0e77 on stream testFailoverScopegepsh/testFailoverStreamJscoF
scrambler-slave-1-container-2/controller_2018-01-23.21.log:2018-01-23 06:57:27,304 3965684 [controllerpool-11] DEBUG i.p.c.s.e.r.AbortRequestHandler - Aborting transaction 7a16b5ca-14b4-4f53-8fe0-a2f2057e0e77 on stream testFailoverScopegepsh/testFailoverStreamJscoF
```


4) The additional logging indicates the client read @ 06:36:12 never completed.

``` scrambler-slave-1-container-2/controller_2018-01-23.21.log:2018-01-23 06:36:12,963 2691343 [EventProcessor[abortStreamReaders:0]] DEBUG i.p.c.e.impl.EventProcessorCell - ==> Attempt to read event EventProcessor[abortStreamReaders:0]
scrambler-slave-1-container-2/controller_2018-01-23.21.log:2018-01-23 06:57:15,383 3953763 [EventProcessor[abortStreamReaders:0]] DEBUG i.p.c.e.impl.EventProcessorCell - ==> Attempt to read event EventProcessor[abortStreamReaders:0]
scrambler-slave-1-container-2/controller_2018-01-23.21.log:2018-01-23 06:57:27,043 3965423 [EventProcessor[abortStreamReaders:0]] DEBUG i.p.c.e.impl.EventProcessorCell - ==> Event read EventProcessor[abortStreamReaders:0] is EventReadImpl(eventSequence=Sequence(highOrder=0, lowOrder=307), event=null, position=PositionImpl(ownedSegments=[Segment(scope=_system, streamName=_abortStream, segmentNumber=1), Segment(scope=_system, streamName=_abortStream, segmentNumber=0)]), eventPointer=null, checkpointName=921294f9-3db9-4b41-906e-4c6fcf51323f) 
```
5) 
After the read request @06:36:12, the reader loses connection to _abortStream/0. 
Then it receives
```2018-01-23 06:36:14,809 2693189 [EventProcessor[abortStreamReaders:0]] WARN  i.p.c.s.impl.SegmentInputStreamImpl - Encountered an exception while reading for Segment(scope=_system, streamName=_abortStream, segmentNumber=1)
io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
```
After this there are 12 such attempts to connect to the _abortStream/1 and then there are no logs related to connection attempts.
```
2018-01-23 06:38:38,817 2837197 [epollEventLoopGroup-2-6] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Closing connection to segment Segment(scope=_system, streamName=_abortStream, segmentNumber=0) with exception: {}
io.pravega.shared.protocol.netty.ConnectionFailedException: null
    at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl$ResponseProcessor.connectionDropped(AsyncSegmentInputStreamImpl.java:58)
```

Interestingly this issue was never seen once the controller disabled automatic checkpointing.

[controller_2018-01-23.21.log.zip](https://github.com/pravega/pravega/files/1662530/controller_2018-01-23.21.log.zip)

**Problem location**
Reader.

**Suggestions for an improvement**
Investigate further and fix the issue.

I can also see from the log there was a RetriesExhaustedException.
```
2018-01-30 07:37:14,246 2306619 [EventProcessor[abortStreamReaders:0]] WARN  i.p.c.s.impl.SegmentInputStreamImpl - Encountered an exception while reading for Segment(scope=_system, streamName=_abortStream, segmentNumber=1)
io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
        at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$null$6(Retry.java:266)
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:870)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:852)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.failAllInflight(AsyncSegmentInputStreamImpl.java:228)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.closeConnection(AsyncSegmentInputStreamImpl.java:200)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.access$100(AsyncSegmentInputStreamImpl.java:39)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl$ResponseProcessor.connectionDropped(AsyncSegmentInputStreamImpl.java:58)
        at io.pravega.client.netty.impl.ClientConnectionInboundHandler.channelUnregistered(ClientConnectionInboundHandler.java:86)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:160)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelUnregistered(DefaultChannelPipeline.java:1337)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:167)
        at io.netty.channel.DefaultChannelPipeline.fireChannelUnregistered(DefaultChannelPipeline.java:835)
        at io.netty.channel.AbstractChannel$AbstractUnsafe$7.run(AbstractChannel.java:771)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:309)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
        at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:326)
        at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:338)
        at java.util.concurrent.CompletableFuture.uniRelay(CompletableFuture.java:911)
        at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:899)
        ... 33 common frames omitted
Caused by: io.pravega.shared.protocol.netty.ConnectionFailedException: null
        ... 28 common frames omitted
```
This does explain why the stack trace was in the state in was in at the time of the dump.

--- a/client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java
+++ b/client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java
@@ -18,7 +18,6 @@ import io.pravega.shared.protocol.netty.WireCommandType;
 import io.pravega.shared.protocol.netty.WireCommands;
 import io.pravega.shared.protocol.netty.WireCommands.SegmentRead;
 import java.nio.ByteBuffer;
-import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
 import javax.annotation.concurrent.GuardedBy;
 import lombok.Synchronized;

--- a/client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java
+++ b/client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java
@@ -74,12 +74,7 @@ class SegmentInputStreamImpl implements SegmentInputStream {
          */
         this.readLength = Math.min(DEFAULT_READ_LENGTH, bufferSize);
         this.buffer = new CircularBuffer(Math.max(bufferSize, readLength + 1));
-
-        try {
-            issueRequestIfNeeded();
-        } catch (SegmentTruncatedException e) {
-            log.warn("{} is already truncated at it's initial offset of {}", asyncInput, offset);
-        }
+        issueRequestIfNeeded();
     }
 
     @Override
@@ -172,8 +167,10 @@ class SegmentInputStreamImpl implements SegmentInputStream {
         try {
             segmentRead = outstandingRequest.join();
         } catch (Exception e) {
+            outstandingRequest = null;
             if (Exceptions.unwrap(e) instanceof SegmentTruncatedException) {
                 receivedTruncated = true;
+                throw new SegmentTruncatedException(e);
             }
             throw e;
         }
@@ -200,23 +197,9 @@ class SegmentInputStreamImpl implements SegmentInputStream {
     /**
      * Issues a request if there is enough room for another request, and we aren't already waiting on one
      */
-    private void issueRequestIfNeeded() throws SegmentTruncatedException {
-        if (!receivedEndOfSegment && !receivedTruncated && buffer.capacityAvailable() >= readLength) {
-            if (outstandingRequest == null) {
-                outstandingRequest = asyncInput.read(offset + buffer.dataAvailable(), readLength);
-            } else if (outstandingRequest.isCompletedExceptionally()) {
-                Throwable e = Futures.getException(outstandingRequest);
-                Throwable realException = Exceptions.unwrap(e);
-                if (realException instanceof SegmentTruncatedException) {
-                    receivedTruncated = true;
-                    throw new SegmentTruncatedException(e);
-                }
-                if (!(realException instanceof Error || realException instanceof InterruptedException
-                        || realException instanceof CancellationException)) {
-                    log.warn("Encountered an exception while reading for " + asyncInput.getSegmentId(), e);
-                    outstandingRequest = asyncInput.read(offset + buffer.dataAvailable(), readLength);
-                }
-            }
+    private void issueRequestIfNeeded() {
+        if (!receivedEndOfSegment && !receivedTruncated && buffer.capacityAvailable() >= readLength && outstandingRequest == null) {
+            outstandingRequest = asyncInput.read(offset + buffer.dataAvailable(), readLength);
         }
     }
 

--- a/client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java
+++ b/client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java
@@ -168,10 +168,15 @@ class SegmentInputStreamImpl implements SegmentInputStream {
     }
 
     private void handleRequest() throws SegmentTruncatedException {
-        if (outstandingRequest.isCompletedExceptionally()) {
-            issueRequestIfNeeded();
+        SegmentRead segmentRead;
+        try {
+            segmentRead = outstandingRequest.join();
+        } catch (Exception e) {
+            if (Exceptions.unwrap(e) instanceof SegmentTruncatedException) {
+                receivedTruncated = true;
+            }
+            throw e;
         }
-        SegmentRead segmentRead = outstandingRequest.join();
         verifyIsAtCorrectOffset(segmentRead);
         if (segmentRead.getData().hasRemaining()) {
             buffer.fill(segmentRead.getData());

--- a/client/src/test/java/io/pravega/client/segment/impl/SegmentInputStreamTest.java
+++ b/client/src/test/java/io/pravega/client/segment/impl/SegmentInputStreamTest.java
@@ -149,6 +149,7 @@ public class SegmentInputStreamTest {
         ByteBuffer read = stream.read(10);
         assertNull(read);
         fakeNetwork.completeExceptionally(1, new ConnectionFailedException());
+        AssertExtensions.assertThrows(ConnectionFailedException.class, () -> stream.read());
         stream.read(10);
         assertNull(read);
         read = stream.read(10);

--- a/client/src/test/java/io/pravega/client/segment/impl/SegmentInputStreamTest.java
+++ b/client/src/test/java/io/pravega/client/segment/impl/SegmentInputStreamTest.java
@@ -136,7 +136,7 @@ public class SegmentInputStreamTest {
         assertEquals(ByteBuffer.wrap(data), read);
     }
 
-    @Test(timeout=10000)
+    @Test(timeout = 10000)
     public void testTimeout() throws EndOfSegmentException, SegmentTruncatedException {
         byte[] data = new byte[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
         ByteBuffer wireData = createEventFromData(data);

--- a/client/src/test/java/io/pravega/client/segment/impl/SegmentInputStreamTest.java
+++ b/client/src/test/java/io/pravega/client/segment/impl/SegmentInputStreamTest.java
@@ -27,6 +27,7 @@ import org.junit.Test;
 import static io.pravega.test.common.Async.testBlocking;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 
 public class SegmentInputStreamTest {
@@ -135,18 +136,40 @@ public class SegmentInputStreamTest {
         assertEquals(ByteBuffer.wrap(data), read);
     }
 
+    @Test(timeout=10000)
+    public void testTimeout() throws EndOfSegmentException, SegmentTruncatedException {
+        byte[] data = new byte[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+        ByteBuffer wireData = createEventFromData(data);
+        TestAsyncSegmentInputStream fakeNetwork = new TestAsyncSegmentInputStream(segment, 7);
+        @Cleanup
+        SegmentInputStreamImpl stream = new SegmentInputStreamImpl(fakeNetwork, 0);
+
+        testBlocking(() -> stream.read(),
+                     () -> fakeNetwork.complete(0, new WireCommands.SegmentRead(segment.getScopedName(), 0, false, false, wireData.slice())));
+        ByteBuffer read = stream.read(10);
+        assertNull(read);
+        fakeNetwork.completeExceptionally(1, new ConnectionFailedException());
+        stream.read(10);
+        assertNull(read);
+        read = stream.read(10);
+        assertNull(read);
+    }
+    
+    
     @Test
     public void testExceptionRecovery() throws EndOfSegmentException, SegmentTruncatedException {
         byte[] data = new byte[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
         ByteBuffer wireData = createEventFromData(data);
-        TestAsyncSegmentInputStream fakeNetwork = new TestAsyncSegmentInputStream(segment, 6);
+        TestAsyncSegmentInputStream fakeNetwork = new TestAsyncSegmentInputStream(segment, 7);
         @Cleanup
         SegmentInputStreamImpl stream = new SegmentInputStreamImpl(fakeNetwork, 0);
         fakeNetwork.complete(0, new WireCommands.SegmentRead(segment.getScopedName(), 0, false, false, ByteBufferUtils.slice(wireData, 0, 2)));
         fakeNetwork.completeExceptionally(1, new ConnectionFailedException());
-        fakeNetwork.complete(2, new WireCommands.SegmentRead(segment.getScopedName(), 2, false, false, ByteBufferUtils.slice(wireData, 2, 7)));
-        fakeNetwork.complete(3, new WireCommands.SegmentRead(segment.getScopedName(), 9, false, false, ByteBufferUtils.slice(wireData, 9, 2)));
-        fakeNetwork.complete(4, new WireCommands.SegmentRead(segment.getScopedName(), 11, false, false, ByteBufferUtils.slice(wireData, 11, wireData.capacity() - 11)));
+        fakeNetwork.complete(2, new WireCommands.SegmentRead(segment.getScopedName(), 0, false, false, ByteBufferUtils.slice(wireData, 0, 2)));
+        fakeNetwork.complete(3, new WireCommands.SegmentRead(segment.getScopedName(), 2, false, false, ByteBufferUtils.slice(wireData, 2, 7)));
+        fakeNetwork.complete(4, new WireCommands.SegmentRead(segment.getScopedName(), 9, false, false, ByteBufferUtils.slice(wireData, 9, 2)));
+        fakeNetwork.complete(5, new WireCommands.SegmentRead(segment.getScopedName(), 11, false, false, ByteBufferUtils.slice(wireData, 11, wireData.capacity() - 11)));
+        AssertExtensions.assertThrows(ConnectionFailedException.class, () -> stream.read());
         ByteBuffer read = stream.read();
         assertEquals(ByteBuffer.wrap(data), read);
     }

